global-hooks/deckhouse-config/startup_sync_test.go

```diff
@@ -0,0 +1,423 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"fmt"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gbytes"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8syaml "sigs.k8s.io/yaml"
+
+	d8config "github.com/deckhouse/deckhouse/go_lib/deckhouse-config"
+	"github.com/deckhouse/deckhouse/go_lib/deckhouse-config/conversion"
+	module_manager "github.com/deckhouse/deckhouse/go_lib/deckhouse-config/module-manager"
+	. "github.com/deckhouse/deckhouse/testing/hooks"
+)
+
+const testGlobalHooksDir = "testdata/test-startup-sync/global-hooks"
+const testModulesDir = "testdata/test-startup-sync/modules"
+
+var _ = Describe("Global hooks :: deckhouse-config :: migrate", func() {
+	f := HookExecutionConfigInit(`{"global": {"discovery": {}}}`, `{}`)
+	// Emulate ensure_crd hook.
+	f.RegisterCRD("deckhouse.io", "v1", "ModuleConfig", false)
+
+	BeforeEach(func() {
+		// Load addon-operator with 3 modules: deckhouse, cert-manager and prometheus.
+		mm, initErr := module_manager.InitBasic(testGlobalHooksDir, testModulesDir)
+		d8config.InitService(mm)
+		Expect(initErr).ShouldNot(HaveOccurred(), "should init module manager: %s", initErr)
+	})
+
+	Context("Phase 1. Migrate deployment/deckhouse to generated ConfigMap", func() {
+
+		BeforeEach(func() {
+			// Prepare non-migrated deckhouse.
+			_ = os.Setenv("ADDON_OPERATOR_CONFIG_MAP", "deckhouse")
+			depl := d8Deployment(d8config.DeckhouseConfigMapName)
+			cm := d8ConfigMap(d8config.DeckhouseConfigMapName, `
+global: |
+  param1: val1
+  param2: val2
+deckhouse: |
+  p1: val1
+certManager: |
+  param1: val1
+certManagerEnabled: "true"
+`)
+			f.KubeStateSet(depl + cm)
+
+			f.BindingContexts.Set(f.GenerateOnStartupContext())
+			f.RunHook()
+		})
+
+		It("should create generated cm and update deployment", func() {
+			Expect(f).To(ExecuteSuccessfully())
+
+			// Check generated ConfigMap.
+			generatedCM := f.KubernetesResource("ConfigMap", "d8-system", d8config.GeneratedConfigMapName)
+			Expect(generatedCM.Exists()).Should(BeTrue())
+			annotationJSON := fmt.Sprintf(`{"%s":"true"}`, shouldMigrateAnnotation)
+			Expect(generatedCM.Field("metadata.annotations").String()).Should(MatchJSON(annotationJSON))
+			Expect(generatedCM.Field("data.global").String()).Should(ContainSubstring("param1: val1"))
+			Expect(generatedCM.Field("data.deckhouse").Exists()).Should(BeTrue())
+			Expect(generatedCM.Field("data.certManager").Exists()).Should(BeTrue())
```

Comment:
Maybe we need to add values comparing with existing data fields?

Reply:
Actually it is not important, as ConfigMap is just copied here without transformations. I've added more checks anyway.